internal/mesh/internal/controllers/gatewayproxy/controller.go

```diff
@@ -85,14 +87,119 @@ func (r *reconciler) Reconcile(ctx context.Context, rt controller.Runtime, req c
+	proxyStateTemplate, err := dataFetcher.FetchProxyStateTemplate(ctx, req.ID)
+	if err != nil {
+		rt.Logger.Error("error reading proxy state template", "error", err)
+		return nil
+	}
+
+	if proxyStateTemplate == nil {
+		req.ID.Uid = ""
+		rt.Logger.Trace("proxy state template for this gateway doesn't yet exist; generating a new one")
+	}
+
+	gwID := &pbresource.ID{
+		Name:    workload.Data.Identity,
+		Type:    pbmesh.APIGatewayType,
+		Tenancy: workload.Id.Tenancy,
+	}
+
+	apiGateway, err := dataFetcher.FetchAPIGateway(ctx, gwID)
+	if err != nil {
+		rt.Logger.Error("error reading the associated api gateway", "error", err)
+		return err
+	}
+
+	if apiGateway == nil {
+		rt.Logger.Trace("api gateway doesn't exist; skipping reconciliation", "apiGatewayID", gwID)
+		return nil
+	}
+
+	allTCPRoutes, err := dataFetcher.FetchAllTCPRoutes(ctx, req.ID.Tenancy)
+	if err != nil {
+		rt.Logger.Error("error reading the associated tcp routes", "error", err)
+		return err
+	}
+
+	if len(allTCPRoutes) == 0 {
+		rt.Logger.Trace("no tcp routes found for this gateway", "apiGatewayID", gwID)
+	}
+
+	services := make([]*pbcatalog.Service, 0)
+
+	referencedTCPRoutes := make([]*pbmesh.TCPRoute, 0)
+	for _, tcpRoute := range allTCPRoutes {
+		for _, parentRef := range tcpRoute.Data.ParentRefs {
+			if parentRef.Ref.Name == apiGateway.Id.Name {
+				referencedTCPRoutes = append(referencedTCPRoutes, tcpRoute.Data)
+				for _, rule := range tcpRoute.Data.Rules {
+					for _, backendRef := range rule.BackendRefs {
+						service, err := dataFetcher.FetchService(ctx, resource.IDFromReference(backendRef.BackendRef.Ref))
+						if err != nil {
+							rt.Logger.Error("error reading the associated service", "error", err)
+							return err
+						}
+
+						if service == nil {
+							rt.Logger.Error("service was nil", "serviceID", resource.IDFromReference(backendRef.BackendRef.Ref))
+							return nil
+						}
+						services = append(services, service.Data)
+					}
+				}
+			}
+		}
+	}
+
+	trustDomain, err := r.getTrustDomain()
+	if err != nil {
+		rt.Logger.Error("error fetching trust domain to compute proxy state template", "error", err)
+		return err
+	}
+
+	gw := apiGateway.Data
+
+	newPST := builder.NewAPIGWProxyStateTemplateBuilder(workload, services, referencedTCPRoutes, gw, rt.Logger, dataFetcher, r.dc, trustDomain).Build()
+
+	proxyTemplateData, err := anypb.New(newPST)
+	if err != nil {
+		rt.Logger.Error("error creating proxy state template data", "error", err)
+		return err
+	}
+	rt.Logger.Trace("updating proxy state template")
+
+	// If we're not creating a new PST and the generated one matches the existing one, nothing to do
+	if proxyStateTemplate != nil && proto.Equal(proxyStateTemplate.Data, newPST) {
+		rt.Logger.Trace("no changes to existing proxy state template")
+		return nil
+	}
```

Comment:
#nit it seems it would make more sense to log `updating proxy state template` _after_ verifying that the PST doesn't currently match desired state

Reply:
can do, I think we do this same pattern for mesh gateways so I'll modify it there as well